### PR TITLE
Add VU check for sampled image descriptors

### LINKED ISSUE #551: Missing image layout validation on draw
I came across a bug recently where an image was not transitioned to the right layout, and the validation layers didn't catch this. In short, the application does the following:

```
1. Transition image 1 to COLOR_ATTACHMENT_OPTIMAL
2. Transition image 2 to SHADER_READ_ONLY_OPTIMAL
3. Draw call (that copies image 2 into image 1 with some transformations)
4. Transition image 3 to COLOR_ATTACHMENT_OPTIMAL
5. Transition image 1 to SHADER_READ_ONLY_OPTIMAL
6. Draw call (that copies image 1 into image 3 for CPU read back)
7. Transition image 3 to TRANSFER_SRC_OPTIMAL
8. Copy image 3 to buffer
```

The bug was that step 5 was missing, i.e. image 1 was in COLOR_ATTACHMENT_OPTIMAL but was being read by a fragment shader. In effect, the color cache was not flushed and the changes in step 3 were not visible to step 6.

It looks like the validation layers don't validate the image layouts on draw call. As a test, I change step 5 to transition to TRANSFER_SRC_OPTIMAL with no complaint from the layers either.

[This page](https://vulkan.lunarg.com/doc/view/1.0.30.0/windows/validation_layer_details.html) indicates the image layouts are validated only in the following functions (though the page is dated):

```
vkCreateRenderPass
vkMapMemory
vkQueuePresentKHR
vkQueueSubmit
vkCmdCopyImage
vkCmdCopyImageToBuffer
vkCmdWaitEvents
VkCmdPipelineBarrier
```

Is there a reason why they are not validated on draw/dispatch calls too?

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Implement check and unit test for VUID-VkWriteDescriptorSet-descriptorType-01403 which requires that descriptor updates for sampled images have specific image layout values.  Note that the language of the spec is being updated to refer to the body text which gives the full set of (extension specific) valid image layout values.

Addtionally
* fixed existing issue in imageLayout tracking in test framework
* refactored/simplified test case which would get the ...-01403 unit test.

Fixes #551

### CODE REVIEW

File: layers/descriptor_sets.cpp [hunk 1]
```diff
@@ -1497,6 +1497,37 @@ bool cvdescriptorset::ValidateImageUpdate(VkImageView image_view, VkImageLayout
         *error_msg = error_str.str();
         return false;
     }
+
+    // Test that the layout is compatible with the descriptorType
```
Comment (jzulauf-lunarg) [inline layers/descriptor_sets.cpp]: The logic of which layouts are enable for a given set of extension matches the if defs in the spec, and the vector is constructed s.t. we don't have to do this logic tree in a complex boolean expression (and that we only do it *once* in code)

File: layers/descriptor_sets.cpp [hunk 2]
```diff
@@ -1497,6 +1497,37 @@ bool cvdescriptorset::ValidateImageUpdate(VkImageView image_view, VkImageLayout
...
+    }
+    if (std::find(valid_layouts.cbegin(), valid_layouts.cend(), image_layout) == valid_layouts.cend()) {
+        *error_code = "VUID-VkWriteDescriptorSet-descriptorType-01403";
+        std::stringstream error_str;
+        error_str << "Descriptor update with descriptorType " << string_VkDescriptorType(type)
+                  << " is being updated with invalid imageLayout " << string_VkImageLayout(image_layout)
+                  << ". Allowed layouts are: ";
+
+        for (size_t i = 0; i < valid_layouts.size(); i++) {
+            if (i > 0) {
```
Comment (jzulauf-lunarg) [inline layers/descriptor_sets.cpp]: strangely, there is no "oxford comma" join utility... :)

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -2277,7 +2277,7 @@ TEST_F(VkLayerTest, GpuValidationArrayOOB) {
 
     VkDescriptorImageInfo image_info[6] = {};
     for (int i = 0; i < 6; i++) {
-        image_info[i] = texture.m_imageInfo;
+        image_info[i] = texture.DescriptorImageInfo();
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: this is part of the framework cleanup for correct image layout tracking.  The former code had a "man with two watches" problem.  It tracked to different DescriptorImageInfo structs, partially updating each (and thus making part of each stale)

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -8493,73 +8493,29 @@ TEST_F(VkLayerTest, WriteDescriptorSetIntegrityCheck) {
...
-
-    VkDescriptorSetAllocateInfo alloc_info = {};
-    alloc_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO;
-    alloc_info.descriptorSetCount = 1;
-    alloc_info.descriptorPool = ds_pool;
-    alloc_info.pSetLayouts = &ds_layout.handle();
-    VkDescriptorSet descriptorSet;
-    err = vkAllocateDescriptorSets(m_device->device(), &alloc_info, &descriptorSet);
-    ASSERT_VK_SUCCESS(err);
+    OneOffDescriptorSet::Bindings bindings = {
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: Refactoring to OneOffDescriptorSet saves about 50 lines of implementation and makes adding another descriptor for the added test a one line change.

File: tests/vkrenderframework.cpp [hunk 5]
```diff
@@ -1269,7 +1267,7 @@ VkTextureObj::VkTextureObj(VkDeviceObj *device, uint32_t *colors) : VkImageObj(d
     /* create image view */
     view.image = handle();
     m_textureView.init(*m_device, view);
-    m_imageInfo.imageView = m_textureView.handle();
+    m_descriptorImageInfo.imageView = m_textureView.handle();
```
Comment (jzulauf-lunarg) [inline tests/vkrenderframework.cpp]: We use the VkImageObj::m_descriptorImageInfo (from the base class) as the one place we track this information

File: tests/vkrenderframework.h [hunk 6]
```diff
@@ -326,7 +326,7 @@ class VkTextureObj : public VkImageObj {
    public:
     VkTextureObj(VkDeviceObj *device, uint32_t *colors = NULL);
 
-    VkDescriptorImageInfo m_imageInfo;
+    const VkDescriptorImageInfo &DescriptorImageInfo() const { return m_descriptorImageInfo; }
```
Comment (jzulauf-lunarg) [inline tests/vkrenderframework.h]: and replace a public member with an accessor to the protected data.

**Comment by jzulauf-lunarg** [comment]
@ShabbyX  -- this PR should provide for the needed validation message at WriteDescriptor time.  If the image is in the incorrect layout (i.e. one that doesn't match the validated imageLayout in the descriptor) existing draw-time validation (at record or queue submit time, depending) should catch that.  If the imageLayout matches the incorrect one in the descriptor, that won't generate a draw-time error, but validation has *already* reported that the descriptor imageLayout is incorrect for this type.

**Comment by ShabbyX** [comment]
Thanks for doing this. Were you successful in incurring a validation error in the example I provided?

File: layers/descriptor_sets.cpp (hunk 2 above)
Comment (ShabbyX) [reply layers/descriptor_sets.cpp]: What I usually do is `error_str << (i > 0 ? ", " : "") << etc;`, though adding `" and "` makes it a bit less elegant.

File: tests/layer_validation_tests.cpp [hunk 7]
```diff
@@ -8623,10 +8579,28 @@ TEST_F(VkLayerTest, WriteDescriptorSetIntegrityCheck) {
...
+    VkImageObj image(m_device);
+    const VkFormat tex_format = VK_FORMAT_B8G8R8A8_UNORM;
+    image.Init(32, 32, 1, tex_format, VK_IMAGE_USAGE_SAMPLED_BIT, VK_IMAGE_TILING_OPTIMAL, 0);
+    ASSERT_TRUE(image.initialized());
+
+    // Attmept write with incorrect
+    imageInfo.sampler = VK_NULL_HANDLE;
+    imageInfo.imageView = image.targetView(tex_format);
+    imageInfo.imageLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+    ;
```
Comment (ShabbyX) [inline tests/layer_validation_tests.cpp]: Stray `;`

**Comment by jzulauf-lunarg** [comment]
@ShabbyX -- yes I was able to recreate your error.  The unit test emulates the failure, and generates a Validation error message at UpdateDescriptors time.  However, I broke the system I had angle set up on before I finished this change.  If you want to pull and test the PR against your original case that would be appreciated.

**Comment by ShabbyX** [comment]
Will do.

**Comment by ShabbyX** [comment]
Unfortunately, we're blocked on a few issues preventing us from rolling the validation layers forward, and cherry-picking this back doesn't build cleanly. I'll try to remember to test this after those issues are resolved.

**Comment by jzulauf-lunarg** [comment]
@mikes-lunarg @tobine  -- ready and clean

File: tests/layer_validation_tests.cpp (hunk 4 above)
Comment (mikes-lunarg) [reply tests/layer_validation_tests.cpp]: Nice!

File: layers/descriptor_sets.cpp [hunk 8]
```diff
@@ -1497,6 +1497,37 @@ bool cvdescriptorset::ValidateImageUpdate(VkImageView image_view, VkImageLayout
         *error_msg = error_str.str();
         return false;
     }
+
+    // Test that the layout is compatible with the descriptorType
+    std::vector<VkImageLayout> valid_layouts = {VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL,
```
[the remark this answers is not in the document]
Comment (jzulauf-lunarg) [reply layers/descriptor_sets.cpp]: That's a better approach (the other thought was to build this a CreateDevice time), but a static array would be cleaner.

**Comment by jzulauf-lunarg** [comment]
@mikes-lunarg  -- check the table/message rework?

File: layers/descriptor_sets.cpp [hunk 9]
```diff
@@ -1497,6 +1498,46 @@ bool cvdescriptorset::ValidateImageUpdate(VkImageView image_view, VkImageLayout
...
+            {//  Note double brace req'd for aggregate initialization
+             {VK_IMAGE_LAYOUT_SHARED_PRESENT_KHR, &DeviceExtensions::vk_khr_shared_presentable_image},
+             {VK_IMAGE_LAYOUT_DEPTH_READ_ONLY_STENCIL_ATTACHMENT_OPTIMAL, &DeviceExtensions::vk_khr_maintenance2},
+             {VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_STENCIL_READ_ONLY_OPTIMAL, &DeviceExtensions::vk_khr_maintenance2}}};
+        auto is_layout = [image_layout, dev_data](const ExtensionLayout &ext_layout) {
+            return dev_data->extensions.*(ext_layout.extension) && (ext_layout.layout == image_layout);
+        };
+
+        bool valid_layout = (std::find(valid_layouts.cbegin(), valid_layouts.cend(), image_layout) != valid_layouts.cend()) ||
+                            std::any_of(extended_layouts.cbegin(), extended_layouts.cend(), is_layout);
```
Comment (mikes-lunarg) [inline layers/descriptor_sets.cpp]: clever!

Comment (jzulauf-lunarg) [reply layers/descriptor_sets.cpp]: @mikes-lunarg any_of is pretty cool and both it and find short-circuit on the "found one" case

**Pull request merged**